xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,251 @@
+                if (oldCdsConfig != null) {
+                  xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+                }
+                CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+              }
+            }
+          };
+        }
+      };
+
+      switchingLoadBalancer.switchTo(fixedCdsConfigBalancer);
+    }
+
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+
+    // clusterWatcher is also updated after switchingLoadBalancer.handleResolvedAddresses
+    cdsConfig = newCdsConfig;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Name resolution error: '%s'", error);
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "CDS load balancer is shutting down");
+
+    switchingLoadBalancer.shutdown();
+    if (xdsClientRef != null) {
+      xdsClientRef.returnObject(xdsClient);
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final LoadBalancer edsBalancer;
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+      this.edsBalancer = lbRegistry.getProvider("xds_experimental").newLoadBalancer(helper);
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate update) {
+      Preconditions.checkArgument(
+          update.getLbPolicy().equals("round_robin"),
+          "The lbPolicy in ClusterUpdate '%s' is not 'round_robin'", update);
+
+      final XdsConfig edsConfig = new XdsConfig(
+          /* balancerName = */ null,
+          new LbConfig(update.getLbPolicy(), ImmutableMap.<String, Object>of()),
+          /* fallbackPolicy = */ null,
+          /* edsServiceName = */ update.getEdsServiceName(),
+          /* lrsServerName = */ update.getLrsServerName());
+
+      LoadStatsStore loadStatsStore = new LoadStatsStoreImpl();
+      if (update.isEnableLrs()) {
+        xdsClient.reportClientStats(
+            update.getClusterName(), update.getLrsServerName(), loadStatsStore);
+      }
+      edsBalancer.handleResolvedAddresses(
+          resolvedAddresses.toBuilder()
+              .setAttributes(
+                  resolvedAddresses.getAttributes().toBuilder()
+                      .discard(ATTR_LOAD_BALANCING_CONFIG)
+                      .set(XdsAttributes.LOAD_STATS_STORE_REF, loadStatsStore)
+                      .build())
+              .setLoadBalancingPolicyConfig(edsConfig)
+          .build());
+    }
+
+    @Override
+    public void onError(Status error) {
+      helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+  }
+
+  static final class CdsConfig {
```

Comment:
private?

Reply:
This `CdsLoadBalancer` class itself does not create an instance of it, so some other class must create it. Basically the `CdsLoadBalancerProvider` (not added yet) will implement `LoadBalancerProvider#parseLoadBalancingPolicyConfig()` method which can create a `ResolvedAddresses` with its `ResolvedAddresses.getLoadBalancingPolicyConfig()` method returning a `CdsConfig` object.